comp/core/settings/component.go

```diff
@@ -0,0 +1,77 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+// Package settings defines the interface for the component that manage settings that can be changed at runtime
+package settings
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/config/model"
+	"go.uber.org/fx"
+)
+
+// team: agent-shared-components
+
+// SettingNotFoundError is used to warn about non existing/not registered runtime setting
+type SettingNotFoundError struct {
+	Name string
+}
+
+func (e *SettingNotFoundError) Error() string {
+	return fmt.Sprintf("setting %s not found", e.Name)
+}
+
+// RuntimeSettingResponse is used to communicate settings config
+type RuntimeSettingResponse struct {
+	Description string
+	Hidden      bool
+}
+
+// Component is the component type.
+type Component interface {
+	// RuntimeSettings returns the configurable settings
+	RuntimeSettings() map[string]RuntimeSetting
+	// GetRuntimeSetting returns the value of a runtime configurable setting
+	GetRuntimeSetting(setting string) (interface{}, error)
+	// SetRuntimeSetting changes the value of a runtime configurable setting
+	SetRuntimeSetting(setting string, value interface{}, source model.Source) error
+
+	// API related functions
+	// Todo: (Components) Remove these functions once we can register routes using FX value groups
+	// GetFullConfig returns the full config
+	GetFullConfig(cfg config.Config, namespaces ...string) http.HandlerFunc
+	// GetValue allows to retrive the runtime setting
+	GetValue(setting string, w http.ResponseWriter, r *http.Request)
+	// SetValue allows to modify the runtime setting
+	SetValue(setting string, w http.ResponseWriter, r *http.Request)
+	// ListConfigurable returns the list of configurable setting at runtime
+	ListConfigurable(w http.ResponseWriter, r *http.Request)
+}
+
+// RuntimeSetting represents a setting that can be changed and read at runtime.
+type RuntimeSetting interface {
+	Get() (interface{}, error)
+	Set(v interface{}, source model.Source) error
+	Name() string
+	Description() string
+	Hidden() bool
+}
+
+// RuntimeSettingProvider stores the Provider instance
+type RuntimeSettingProvider struct {
+	fx.Out
+
+	Setting RuntimeSetting `group:"runtime_setting"`
```

Comment:
If some component wants to declare multiple settings that can be modified at Runtime, does that mean it needs to provide multiple of these? The interface does not pass a name to `Get` or `Set` so it seems like it only represents one setting.

Alternatively, maybe this Provider struct contains `Settings map[name]RuntimeSetting`?

Reply:
That is a great point that I think not only applies to this component but any component that uses Fx value groups. 

For this first iteration of the runtime settings component not a single component pass via Fx value groups rather via the `fx.Provide` in the command: 

```go 
fx.Provide(func() runtimesettings.RuntimeSettingProvider {
	return runtimesettings.NewRuntimeSettingProvider(commonsettings.NewLogLevelRuntimeSetting())
}),
fx.Provide(func(debug dogstatsddebug.Component) runtimesettings.RuntimeSettingProvider {
	return runtimesettings.NewRuntimeSettingProvider(settings.NewDsdStatsRuntimeSetting(debug))
}),
```

I think is a good pattern to have the opportunity to pass arguments via Fx value groups because it enables not just via `fx.Provide` but via other components. Still, it might be overkill we could create a single settings Params that has the signature you mention `map[name]RuntimeSetting`

Reply:
@dustmop I changed so rather than using value groups we pass the settings via a params `type Settings map[name]RuntimeSetting` https://github.com/DataDog/datadog-agent/pull/24292/commits/d1e03c9852199f2eb5ae86aac7a6d03051039239